From 7b98eb0d56da682aefa6a15fbd58f2932790bb0b Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Thu, 22 Aug 2013 20:21:41 +0100
Subject: [PATCH 174/202] TV: Cache the last sourceid found in
 SourceUtil::GetSourceName

This significantly reduces dbase lookups during cursor movements
in GuideGrid.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythtv/libs/libmythtv/sourceutil.cpp |   24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/mythtv/libs/libmythtv/sourceutil.cpp b/mythtv/libs/libmythtv/sourceutil.cpp
index b671dcf..79dd5a7 100644
--- a/mythtv/libs/libmythtv/sourceutil.cpp
+++ b/mythtv/libs/libmythtv/sourceutil.cpp
@@ -2,6 +2,9 @@
 
 // Qt headers
 #include <QRegExp>
+#include <QMap>
+#include <QMutex>
+#include <QMutexLocker>
 
 // MythTV headers
 #include "sourceutil.h"
@@ -41,8 +44,20 @@ bool SourceUtil::HasDigitalChannel(uint sourceid)
     return false;
 }
 
+// static
 QString SourceUtil::GetSourceName(uint sourceid)
 {
+    static QMutex s_mutex;
+    QMutexLocker lock(&s_mutex);
+
+    typedef QMap< uint, QString > SrcNamecache_t;
+    static SrcNamecache_t s_sourceNamecache;
+
+    SrcNamecache_t::const_iterator it = s_sourceNamecache.find(sourceid);
+    if (it != s_sourceNamecache.end())
+        return it.value();
+    lock.unlock();
+
     MSqlQuery query(MSqlQuery::InitCon());
 
     query.prepare(
@@ -51,17 +66,20 @@ QString SourceUtil::GetSourceName(uint sourceid)
         "WHERE sourceid = :SOURCEID");
     query.bindValue(":SOURCEID", sourceid);
 
+    QString sourceName;
     if (!query.exec())
     {
         MythDB::DBError("SourceUtil::GetSourceName()", query);
         return QString::null;
     }
-    else if (!query.next())
+    else if (query.next())
     {
-        return QString::null;
+        sourceName = query.value(0).toString();
     }
 
-    return query.value(0).toString();
+    lock.relock();
+    s_sourceNamecache.insert(sourceid, sourceName);
+    return sourceName;
 }
 
 QString SourceUtil::GetChannelSeparator(uint sourceid)
-- 
1.7.9.5

